lua/luasnip/extras/fmt.lua

```diff
@@ -0,0 +1,159 @@
+        start, end_ = fmt:find(pattern, end_+1)
+        -- if not found, add the remaining part of string and finish
+        if start == nil then
+            add_text(fmt:sub(prev_end + 1))
+            break
+        end
+        -- ignore non-placeholders, i.e. double curly braces replacing them with
+        -- single braces, e.g. `{{2}}` -> `{2}`
+        local before = fmt:sub(start - 1, start)
+        local after = fmt:sub(end_, end_ + 1)
+        local ignored = before == '{{' and after == '}}'
+        if ignored then
+            -- add the previous part of the string
+            add_text(fmt:sub(prev_end + 1, start - 1))
+            -- add what is inside {{}}
+            add_text(fmt:sub(start + 1, end_ - 1))
+            -- move previous end
+            prev_end = end_
+        else
+            -- add the previous part of the string
+            add_text(fmt:sub(prev_end + 1, start - 1))
+            prev_end = end_
+            -- interpret the placeholder
+            local placeholder = fmt:sub(start + 1, end_ - 1)
+            local is_num = placeholder:match('^%d+$')
+            local key
+            if is_num then  -- numbered placeholder
+                last_index = tonumber(placeholder)
+                key = last_index
+            elseif placeholder == '' then  -- auto-numbered placeholder
+                key = last_index + 1
+                last_index = key
+            else  -- named placeholder
+                key = placeholder
+            end
+            add_arg(key)
+        end
+        -- end
+    end
+
+    return elements
+end
+
+-- Find the largest common indent in a list of lines and remove it.
+local function remove_common_indent(lines)
+    local max_indent = 100000
+    for _, line in ipairs(lines) do
+        -- ignore lines with whitespace only
+        local match = line:match('^(%s*)%S')
+        if match then
+            max_indent = math.min(max_indent, #match)
+        end
+    end
+    if max_indent > 0 then
+        lines = vim.tbl_map(function(line)
+            return line:sub(max_indent + 1)
+        end, lines)
+    end
+    return lines
+end
+
+-- Use a format string to replace placeholders with nodes.
+-- The format is similar to Python formatting, e.g. `Some {} then {3} and {name}.`
+-- See `replace_placeholders` documentation for details.
+--
+-- Arguments:
+--   str: format string
+--   nodes: snippet node or list of nodes
+--   opts: optional table
+--     trim_empty: boolean, remove whitespace-only first/last lines, default true
+--     dedent: boolean, remove all common indent in `str`, default true
+-- Returns: list of snippet nodes
+local function format_nodes(str, nodes, opts)
+    opts = opts or {}
+    opts.trim_empty = opts.trim_empty ~= nil and opts.trim_empty or true
+    opts.dedent = opts.dedent ~= nil and opts.dedent or true
+
+    -- allow to pass a single node
+    if type(nodes.is_node) == 'function' and nodes.is_node() then
+        nodes = {nodes}
+    end
```

Comment:
There's already a function similar to that [here](https://github.com/L3MON4D3/LuaSnip/blob/master/lua/luasnip/nodes/snippet.lua#L82-L90), I think it would be nice to factor `wrap_nodes` out to util and use it in both cases.

Reply:
Done